From 99d938d6ea32c31457bb68a5a2177c94d13a4854 Mon Sep 17 00:00:00 2001
From: nyanmisaka <nst799610810@gmail.com>
Date: Thu, 29 Feb 2024 16:55:39 +0800
Subject: [PATCH 1/3] fixup! lavf/rkrga: add RKRGA scale, vpp and overlay
 filter

Drain the fifo during filter reset to avoid leaking one fence_fd.

Signed-off-by: nyanmisaka <nst799610810@gmail.com>
---
 libavfilter/rkrga_common.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/libavfilter/rkrga_common.c b/libavfilter/rkrga_common.c
index f141da5b17..72dbe8c29f 100644
--- a/libavfilter/rkrga_common.c
+++ b/libavfilter/rkrga_common.c
@@ -1048,10 +1048,25 @@ static void set_rga_async_frame_lock_status(RGAAsyncFrame *frame, int lock)
         frame->pat->locked = status;
 }
 
+static void rga_drain_fifo(RKRGAContext *r)
+{
+    RGAAsyncFrame aframe;
+
+    while (r->async_fifo && av_fifo_read(r->async_fifo, &aframe, 1) >= 0) {
+        if (imsync(aframe.dst->info.out_fence_fd) != IM_STATUS_SUCCESS)
+            av_log(NULL, AV_LOG_WARNING, "RGA sync failed\n");
+
+        set_rga_async_frame_lock_status(&aframe, 0);
+    }
+}
+
 av_cold int ff_rkrga_close(AVFilterContext *avctx)
 {
     RKRGAContext *r = avctx->priv;
 
+    /* Drain the fifo during filter reset */
+    rga_drain_fifo(r);
+
     clear_frame_list(&r->src_frame_list);
     clear_frame_list(&r->dst_frame_list);
     clear_frame_list(&r->pat_frame_list);
-- 
2.25.1

